# **Problem Statement**

## A global retail company is facing several supply chain challenges, including delayed deliveries, inefficient shipping operations, fluctuating product demand, and inconsistent profitability across different markets. The management team lacks a centralized analytics solution to monitor the end-to-end supply chain process and make informed business decisions.

## As a Data Analyst, the objective of this project is to analyze the company's supply chain data to identify operational inefficiencies, uncover sales and logistics trends, evaluate customer and product performance, and provide actionable insights that improve overall business performance.

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
from google.colab import files
u=files.upload()

Saving DataCoSupplyChainDataset.csv to DataCoSupplyChainDataset.csv


In [ ]:
df=pd.read_csv(r'/content/DataCoSupplyChainDataset.csv',encoding='latin1')
df.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,...,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [ ]:
df1=df.copy()

In [ ]:
df1.shape

(180519, 53)

In [ ]:
df1.isnull().sum()

,0
Type,0
Days for shipping (real),0
Days for shipment (scheduled),0
Benefit per order,0
Sales per customer,0
Delivery Status,0
Late_delivery_risk,0
Category Id,0
Category Name,0
Customer City,0


In [ ]:
df1['Order Zipcode'].value_counts()

,count
Order Zipcode,
10035.0,648
10009.0,550
10024.0,541
94122.0,526
10011.0,463
...,...
92236.0,1
98002.0,1
60477.0,1


In [ ]:
df1.drop(columns=['Product Description'],axis=1,inplace=True)

The Product Description column contained 100% missing values, making it unsuitable for analysis or feature engineering. Since the column provided no usable information and could not be meaningfully imputed, it was removed from the dataset to reduce unnecessary storage and improve data quality.

In [ ]:
df1.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,...,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,...,COMPLETE,NaN,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,...,PENDING,NaN,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,...,CLOSED,NaN,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,...,COMPLETE,NaN,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,...,PENDING_PAYMENT,NaN,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [ ]:
df1.drop(columns=['Order Zipcode'],axis=1,inplace=True)

The Order Zipcode column contained approximately 86% missing values. Since the dataset already includes City, State, Country, Region, Latitude, and Longitude, the ZIP code was considered redundant and removed.

In [ ]:
df1.isnull().sum()

,0
Type,0
Days for shipping (real),0
Days for shipment (scheduled),0
Benefit per order,0
Sales per customer,0
Delivery Status,0
Late_delivery_risk,0
Category Id,0
Category Name,0
Customer City,0


In [ ]:
df1['Customer Lname'].value_counts()

,count
Customer Lname,
Smith,64104
Johnson,989
Brown,909
Williams,869
Jones,859
...,...
Atom,1
Gerginokish,1
Willi,1


In [ ]:
df1['Customer Lname']=df1['Customer Lname'].fillna('Unknown')

Customer name with null values are filled with unknown keyword.Since it possess only less number of null values.

In [ ]:
df1['Customer Zipcode'].value_counts()

,count
Customer Zipcode,
725.0,66770
921.0,337
23455.0,334
957.0,297
79109.0,292
...,...
60636.0,18
89015.0,16
32210.0,15


The Customer Zipcode column contained 3 missing values (approximately 0.002% of the dataset). Since ZIP codes are identifier fields rather than numerical measures, and the missing percentage is negligible, the null values were retained. This has no significant impact on the overall analysis.

In [ ]:
df1.duplicated().sum()

np.int64(0)

There are no duplicates available.

In [ ]:
df1.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,...,Order State,Order Status,Product Card Id,Product Category Id,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,...,Java Occidental,COMPLETE,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,...,Rajastán,PENDING,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,...,Rajastán,CLOSED,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,...,Queensland,COMPLETE,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,...,Queensland,PENDING_PAYMENT,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 51 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  object 
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  object 
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  object 
 9   Customer City                  180519 non-null  object 
 10  Customer Country               180519 non-null  object 
 11  Customer Email                 180519 non-null  object 
 12  Customer Fname                

In [ ]:
df1['shipping date (DateOrders)']=pd.to_datetime(df1['shipping date (DateOrders)'])

In [ ]:
df1['order date (DateOrders)']=pd.to_datetime(df1['order date (DateOrders)'])

Converting datatype of order date and shipping date to datetime which is in object dtype.

In [ ]:
df1['Customer Zipcode']=df1['Customer Zipcode'].astype('Int64')

In [ ]:
df1['Sales'].value_counts()

,count
Sales,
129.990005,22372
399.980011,17325
199.990005,15622
299.980011,13729
179.970001,5016
...,...
999.989990,10
379.959992,10
99.000000,9


In [ ]:
df1['Order Item Discount'].value_counts()

,count
Order Item Discount,
0.000000,10028
6.000000,4589
12.000000,4067
4.000000,3647
8.000000,3626
...,...
50.490002,1
85.489998,1
140.000000,1


In [ ]:
df1['Order Item Product Price'].value_counts()

,count
Order Item Product Price,
59.990002,24820
129.990005,22372
50.000000,21035
49.980000,19298
399.980011,17325
...,...
349.989990,40
329.989990,27
599.989990,21


In [ ]:
money_cols = [
    'Benefit per order',
    'Sales per customer',
    'Order Item Discount',
    'Order Item Product Price',
    'Sales',
    'Order Item Total',
    'Order Profit Per Order',
    'Product Price'
]

df1[money_cols] = df1[money_cols].round(2)

Several monetary columns contained floating-point precision values (e.g., 129.990005) due to the way floating-point numbers are stored in computer systems. Since financial values are typically represented with two decimal places, these columns were rounded to improve readability, reporting consistency, and data presentation without affecting the underlying business calculations.


In [ ]:
df1['Benefit per order'].value_counts()

,count
Benefit per order,
0.00,1177
143.99,199
72.00,194
46.80,188
24.00,181
...,...
-212.41,1
-18.96,1
177.17,1


In [ ]:
df1['Customer FullName']=(df1['Customer Fname'].str.strip()+ ' ' + df1['Customer Lname'].fillna('').str.strip()).str.strip()

In [ ]:
df1.drop(columns=['Customer Fname','Customer Lname'],axis=1,inplace=True)

The dataset stored customer names in separate First Name and Last Name columns. For improved readability and to simplify customer-level analysis, these columns were combined into a single Customer Full Name field. After verifying the new column, the original name columns were removed to eliminate redundancy.

In [ ]:
df['Product Status'].value_counts(dropna=False)

,count
Product Status,
0,180519


In [ ]:
df1.drop(columns=['Product Status'],axis=1,inplace=True)

The Product Status column was removed because it contained only a single constant value (0) across all records. Since the column exhibited no variability, it did not contribute any meaningful information for analysis or predictive insights.

In [ ]:
df1.drop(columns=['Product Image'],axis=1,inplace=True)

The dataset doesnt need an image column which consists of images of the products.

In [ ]:
(df1['Customer Id'] == df1['Order Customer Id']).all()

np.True_

In [ ]:
(df['Category Id'] == df['Product Category Id']).all()

np.True_

In [ ]:
(df['Product Card Id'] == df['Order Item Cardprod Id']).all()

np.True_

In [ ]:
df1.drop(columns=[
    'Order Customer Id',
    'Product Category Id',
    'Order Item Cardprod Id'
], inplace=True)

During data validation, duplicate identifier columns were identified. The values in Order Customer Id, Product Category Id, and Order Item Cardprod Id were found to be identical to Customer Id, Category Id, and Product Card Id, respectively. Since these columns contained redundant information, they were removed to reduce data redundancy and simplify the dataset without affecting analytical integrity.

In [ ]:
df1.drop(columns=['Customer Email','Customer Password'],axis=1, inplace=True)

Customer Email:
The Customer Email column was removed as it contains personally identifiable information (PII) that is not required for business analysis.
Customer Password:
The Customer Password column was removed because it contains sensitive authentication information that has no analytical value and should not be included in analytics datasets.

In [ ]:
df1['Delivery Status'].value_counts()

,count
Delivery Status,
Late delivery,98977
Advance shipping,41592
Shipping on time,32196
Shipping canceled,7754


In [ ]:
df1['Delivery Status']=df1['Delivery Status'].str.strip().str.lower().str.title()

In [ ]:
df1['Delivery Status'].value_counts()

,count
Delivery Status,
Late Delivery,98977
Advance Shipping,41592
Shipping On Time,32196
Shipping Canceled,7754


In [ ]:
df1['Order Status'].value_counts()

,count
Order Status,
COMPLETE,59491
PENDING_PAYMENT,39832
PROCESSING,21902
PENDING,20227
CLOSED,19616
ON_HOLD,9804
SUSPECTED_FRAUD,4062
CANCELED,3692
PAYMENT_REVIEW,1893


In [ ]:
df1['Order Status']=df1['Order Status'].str.strip().str.lower().str.title()

In [ ]:
df1['Order Status'].value_counts()

,count
Order Status,
Complete,59491
Pending_Payment,39832
Processing,21902
Pending,20227
Closed,19616
On_Hold,9804
Suspected_Fraud,4062
Canceled,3692
Payment_Review,1893


In [ ]:
df1['Shipping Mode'].value_counts()

,count
Shipping Mode,
Standard Class,107752
Second Class,35216
First Class,27814
Same Day,9737


In [ ]:
df1['Market'].value_counts()

,count
Market,
LATAM,51594
Europe,50252
Pacific Asia,41260
USCA,25799
Africa,11614


In [ ]:
df1['Market']=df1['Market'].str.strip().str.lower().str.title()

In [ ]:
df1['Market'].value_counts()

,count
Market,
Latam,51594
Europe,50252
Pacific Asia,41260
Usca,25799
Africa,11614


In [ ]:
df1['Order Status']=df1['Order Status'].replace({'Canceled':'Cancelled','Complete':'Completed'})

In [ ]:
df1['Order Status'].value_counts()

,count
Order Status,
Completed,59491
Pending_Payment,39832
Processing,21902
Pending,20227
Closed,19616
On_Hold,9804
Suspected_Fraud,4062
Cancelled,3692
Payment_Review,1893


In [ ]:
df1['Delivery Status']=df1['Delivery Status'].replace({'Shipping Canceled':'Shipping Cancelled'})

In [ ]:
df1['Delivery Status'].value_counts()

,count
Delivery Status,
Late Delivery,98977
Advance Shipping,41592
Shipping On Time,32196
Shipping Cancelled,7754


In [ ]:
df1['Market']=df1['Market'].replace({'Latam':'Latin America','Usca':'Usa'})

In [ ]:
df1['Market'].value_counts()

,count
Market,
Latin America,51594
Europe,50252
Pacific Asia,41260
Usa,25799
Africa,11614


In [ ]:
df['Type'].value_counts(dropna=False)

,count
Type,
DEBIT,69295
TRANSFER,49883
PAYMENT,41725
CASH,19616


In [ ]:
df1['Type']=df1['Type'].str.strip().str.lower().str.title()

Standardized categorical values by replacing inconsistent category names (e.g., Cancelled -- Canceled and Complete -- Completed) to ensure consistency in analysis and reporting.

Checking for Impossible Values

In [ ]:
(df1['Sales'] < 0).sum()

np.int64(0)

In [ ]:
(df1['Order Item Quantity'] <= 0).sum()

np.int64(0)

In [ ]:
(df1['Product Price'] < 0).sum()

np.int64(0)

In [ ]:
(df1['Days for shipping (real)'] < 0).sum()

np.int64(0)

We can see there are no values which are beyond.The values are correct based on the data.

In [ ]:
df1.rename(columns={'order date (DateOrders)':'Order_Date','shipping date (DateOrders)':'Shipping_Date'},inplace=True)

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 43 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   Type                           180519 non-null  object        
 1   Days for shipping (real)       180519 non-null  int64         
 2   Days for shipment (scheduled)  180519 non-null  int64         
 3   Benefit per order              180519 non-null  float64       
 4   Sales per customer             180519 non-null  float64       
 5   Delivery Status                180519 non-null  object        
 6   Late_delivery_risk             180519 non-null  int64         
 7   Category Id                    180519 non-null  int64         
 8   Category Name                  180519 non-null  object        
 9   Customer City                  180519 non-null  object        
 10  Customer Country               180519 non-null  object        
 11  

In [ ]:
df1.isnull().sum().sum()

np.int64(3)

In [ ]:
df1.duplicated().sum()

np.int64(0)

Lastly checking again before saving the file for EDA.

In [ ]:
df1.to_csv('Supply_Chain.csv',index=False)

In [ ]:
files.download('Supply_Chain.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>